java-spiffe-core/src/test/java/io/spiffe/spiffeid/SpiffeIdTest.java

```diff
@@ -1,89 +1,136 @@
+                Arguments.of("spiffe://example.org/pa:th/element:", TrustDomain.of("example.org"), "/pa:th/element:"),
+                Arguments.of("  spiffe://domain.test/path1/path2  ", TrustDomain.of("domain.test"), "/path1/path2"),
+                Arguments.of("spiffe://example.org/9eebccd2-12bf-40a6-b262-65fe0487d453", TrustDomain.of("example.org"), "/9eebccd2-12bf-40a6-b262-65fe0487d453")
         );
     }
 
-    @Test
-    void of_TrustDomainAndPathsWithCaps_ReturnsSpiffeIdNormalized() {
-        val trustDomain = TrustDomain.of("TRuST-DoMAIN.Org");
-
-        val spiffeId = SpiffeId.of(trustDomain, "PATH1", "paTH2");
+    @ParameterizedTest
+    @MethodSource("provideInvalidSpiffeIds")
+    void testParseInvalidSpiffeId(String input, String expected) {
+        SpiffeId result;
+        try {
+            result = SpiffeId.parse(input);
+            assertEquals(expected, result.toString());
+        } catch (Exception e) {
+            assertEquals(expected, e.getMessage());
+        }
+    }
 
-        assertAll("normalized spiffeId",
-                () -> assertEquals("trust-domain.org", spiffeId.getTrustDomain().toString()),
-                () -> assertEquals("/path1/path2", spiffeId.getPath())
+    static Stream<Arguments> provideInvalidSpiffeIds() {
+        return Stream.of(
+                Arguments.of("", "SPIFFE ID cannot be empty"),
+                Arguments.of(null, "SPIFFE ID cannot be empty"),
+                Arguments.of("192.168.2.2:6688", "SPIFFE ID: malformed URI: 192.168.2.2:6688"),
+                Arguments.of("http://domain.test/path/element", "SPIFFE ID: invalid scheme"),
+                Arguments.of("spiffe:///path/element", "SPIFFE ID: trust domain is empty"),
+                Arguments.of("spiffe://domain.test/path/element?query=1", "SPIFFE ID: query is not allowed"),
+                Arguments.of("spiffe://domain.test/path/element?#fragment-1", "SPIFFE ID: fragment is not allowed"),
+                Arguments.of("spiffe://domain.test:8080/path/element", "SPIFFE ID: port is not allowed"),
+                Arguments.of("spiffe://user:password@test.org/path/element", "SPIFFE ID: user info is not allowed"),
+                Arguments.of("spiffe:path/element", "SPIFFE ID: trust domain is empty"),
+                Arguments.of("spiffe:/path/element", "SPIFFE ID: trust domain is empty"),
+                Arguments.of("spiffe://domain.test/path/elem%5uent", "SPIFFE ID: malformed URI: spiffe://domain.test/path/elem%5uent")
         );
     }
 
-    @Test
-    void of_TrustDomainAndPathWithLeadingAndTrailingBlanks_ReturnsSpiffeIdNormalized() {
-        val trustDomain = TrustDomain.of(" trust-domain.org ");
-
-        val spiffeId = SpiffeId.of(trustDomain, " path1 ", " path2 ");
+    @ParameterizedTest
+    @MethodSource("provideValidTrustDomainAndPaths")
+    void testOf(TrustDomain inputTrustDomain, String[] inputPath, SpiffeId expectedSpiffeId) {
+        SpiffeId result;
```

Comment:
this can be combined with the initialization below since it's unused outside the `try`